lms/djangoapps/instructor/views/api.py

```diff
@@ -2680,3 +2683,116 @@ def start_certificate_generation(request, course_id):
+            {'success': False, 'message': error.message, 'data': json.dumps(certificate_white_list)},
+            status=400
+        )
+
+    message = _('Students added to Certificate white list successfully')
+
+    response_payload = {
+        'success': True,
+        'message': message,
+        'data': json.dumps(certificate_white_list)
+    }
+
+    if white_list_student == 'all':
+        # Generate Certificates for all white listed student
+        students = User.objects.filter(
+            certificatewhitelist__course_id=course_key,
+            certificatewhitelist__whitelist=True
+        )
+        instructor_task.api.generate_certificates_for_students(request, course_key, students=students)
+    elif white_list_student == 'new':
+        # Generate Certificates only for newly added white listed student
+        instructor_task.api.generate_certificates_for_students(request, course_key, students=students)
+
+    return JsonResponse(response_payload)
+
+
+def process_certificate_exceptions(data_list, course_key, students):
+    """
+    Validate user data for certificate exceptions, raise ValueError in case of invalid data and create
+    'CertificateWhitelist' record for students in data_list.
+
+    return updated data_list after creating 'CertificateWhitelist' records in db.
+    """
+    users = [data.get('user_name', False) or data.get('user_email', False) for data in data_list]
+
+    if not all(users):
+        # Username and email can not both be empty
+        raise ValueError('Student username/email is required.')
+
+    if len(users) != len(set(users)):
+        # Duplicate Student username/email is not allowed
+        raise ValueError('Duplicate Student Username/password.')
+
+    for data in data_list:
+        user = data.get('user_name', '') or data.get('user_email', '')
+        try:
+            db_user = get_user(user)
+        except ObjectDoesNotExist:
+            raise ValueError('Student (username/email={user}) does not exists'.format(user=user))
+        except MultipleObjectsReturned:
+            raise ValueError('Multiple Students found with username/email={user}'.format(user=user))
+
+        if CertificateWhitelist.objects.filter(user=db_user, whitelist=True).count() > 0:
+            raise ValueError(
+                "Student (username/email={user_id} already in certificate exception  list)".format(user_id=user)
+            )
+
+        certificate_white_list = CertificateWhitelist.objects.create(
+            user=db_user,
+            course_id=course_key,
+            whitelist=True,
+            free_text=data.get('free_text', '')
+        )
+
+        data.update({
+            'id': certificate_white_list.id,
+            'user_email': db_user.email,
+            'user_name': db_user.username,
+            'user_id': db_user.id,
+            'created': certificate_white_list.created.strftime("%A, %B %d, %Y"),
+        })
+
+        students.append(db_user)
+
+    return data_list, students
```

Comment:
@mattdrayer before, students was being passed as an argument to this function and updated in the function, the caller function was responsible for passing an array object as 'students' arguments and use that value.